README.md

####Travaux pratiques dans le cadre du cours Système.

**Pour tester** : 

* Se rendre sur : <http://jebimoh.fr>

* télécharger le fichier d'exemple fourni en lien (sample.txt)

# Jebimoh 

#### Myvaf (Variant Allele Frequency)

Fournit un graphique du pourcentage de variants obtenus pour un échantillon.
Le fichier doit être formaté de la manière suivante :

	Gene	Variant_sur_Génome_g.	Variant_sur_CDS_c.	Variant_sur_protéine_p_c.	VAF_%
	
**Nota** : le séparateur de champs est une tabulation (et une seule).

#### Kinetic

Cinetique de l'évolution clonale pour un ou plusieurs patients
Le fichier doit être formaté de la manière suivante :

	UPN-23;del(5q) %;TP53 p.R282W VAF %;R282W-CCF;TP53 p.R248Q VAF %;R248Q-CCF;TP53 p.R273C VAF %;R273C-CCF

- Un champs d'entête contenant le motif "-CCF" correspond à un patient

- Une ligne correspond à une date

**Nota** : le séparateur de champs est le ";" (point virgule).

### Les objectifs

* Utiliser python (ok)

* Utiliser CGI (ok)

* Créer un formulaire (ok)

* Créer un élément graphique SVG (ok)

* Utiliser Javascript (ok)

En supléments aux fichier du site, il y a le répertoire **other** :

* Les fichiers **sample*.txt** sont d'autre fichiers d'exemple à télécharger pour tester

* Le fichier **jebimoh.conf** contient la configuration d'Apache2


### Prérequis pour le serveur

* python3 avec les modules : 
	* cgi
	* cgitb					(pour le debugage)
	* python3-markdown (pour lire les fichiers markdown)
	* python3-cairosvg (pour transformer des SVG en PNG)



CHANGELOG.md
### Version 0.5
Le 08/12/2105

- Un peu de nettoyage de code
- Contrôle l'extension du fichier avant téléchargement
- Change les fichiers d'exemple

### Version 0.4
Le 05/12/2015

- kinetik : différencie les chemins (pointillés)
- kinetik : remplace les virgules par des points (erreur float)
- kinetik : intègre la première colonne (del5q)
- kinetik : ajoute un cartouche comme deuxième image

### Version 0.3
Le 03/12/2015

- le design des liens est revu
- Le design du formulaire de téléchargement est revu
- Ajoute la page Kinetic
- Règle le problème de l'encodage (non testé) 
- Ajoute une page "A faire"
- Supprime le bouton "télécharger"
- Télécharge maintenant un fichier png (au lieu d'un SVG)

###Version 0.2
Le 26/11/2015

- Les graphiques sont maintenant téléchargeable au format SVG
- un logo est affiché dans le header
- les barres s'affichent dynamiquement
- github s'affiche dans un autre onglet ou fenêtre


 


tests/test.py
#!/usr/bin/env python3
#! -*- coding:utf-8 -*-

import cairosvg

svg_code1 = """<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" 
  "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="5cm" height="4cm" version="1.1"
     xmlns="http://www.w3.org/2000/svg">
  <desc>Four separate rectangles
  </desc>
    <rect x="0.5cm" y="0.5cm" width="2cm" height="1cm"/>
    <rect x="0.5cm" y="2cm" width="1cm" height="1.5cm"/>
    <rect x="3cm" y="0.5cm" width="1.5cm" height="2cm"/>
    <rect x="3.5cm" y="3cm" width="1cm" height="0.5cm"/>

  <!-- Show outline of canvas using 'rect' element -->
  <rect x=".01cm" y=".01cm" width="4.98cm" height="3.98cm"
        fill="none" stroke="blue" stroke-width=".02cm" />

</svg>
"""

svg_code2 = """<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN"
	"http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd">
	<svg width="620px" height="420px" xml:lang="fr"
	xmlns="http://www.w3.org/2000/svg"
	xmlns:xlink="http://www.w3.org/1999/xlink">

		<title>My VAF</title>
	<rect x="0" y="0" width="620" height="420" fill="white" opacity="0.7"/>
	<path d="M 150,10 L 10,10 L 10,410 L 610,410 L 610,10 L 250,10" style="fill:none; stroke:grey;"/>
	<text id="lisere1" x="160" y="15">sample3</text>
	<line x1="95" y1="330" x2="565" y2="330" stroke="grey"/>
	<line x1="100" y1="335" x2="100" y2="55" stroke="grey"/>
	<text x="90" y="335" style="text-anchor:end;">0 %</text>
	<line x1="95" y1="240.0" x2="565" y2="240.0" stroke="#B3B3B3"/>
	<text x="90" y="245.0" style="text-anchor:end; baseline-shift:5;">10 %</text>
	<line x1="95" y1="150.0" x2="565" y2="150.0" stroke="#B3B3B3"/>
	<text x="90" y="155.0" style="text-anchor:end; baseline-shift:5;">20 %</text>
	<line x1="95" y1="60.0" x2="565" y2="60.0" stroke="#B3B3B3"/>
	<text x="90" y="65.0" style="text-anchor:end; baseline-shift:5;">30 %</text>
	<line x1="157.5" y1="330" x2="157.5" y2="123.0" style=" stroke:grey; stroke-width:15 "/>
	<text x="157.5" y="350" style="text-anchor:middle">ATM</text>
	<g fill="grey"><text x="192.5" y="119.0" style="text-anchor:middle" >H111P</text></g>
	<line x1="272.5" y1="330" x2="272.5" y2="276.0" style=" stroke:grey; stroke-width:15 "/>
	<text x="272.5" y="350" style="text-anchor:middle">BIRC3</text>
	<g fill="grey"><text x="307.5" y="272.0" style="text-anchor:middle" >R22P</text></g>
	<line x1="387.5" y1="330" x2="387.5" y2="195.0" style=" stroke:grey; stroke-width:15 "/>
	<text x="387.5" y="350" style="text-anchor:middle">TP53</text>
	<g fill="grey"><text x="422.5" y="191.0" style="text-anchor:middle" >W40S</text></g>
	<line x1="502.5" y1="330" x2="502.5" y2="60.0" style=" stroke:grey; stroke-width:15 "/>
	<text x="502.5" y="350" style="text-anchor:middle">BRAF</text>
	<g fill="grey"><text x="537.5" y="56.0" style="text-anchor:middle" >Y999Z</text></g>
	</svg>
"""

svg_code3 = '<?xml version="1.0" encoding="utf-8"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN" "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd"><svg width="620px" height="420px" xml:lang="fr" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><title>My VAF</title><rect x="0" y="0" width="620" height="420" fill="white" opacity="0.7"/><path d="M 150,10 L 10,10 L 10,410 L 610,410 L 610,10 L 250,10" style="fill:none; stroke:grey;"/><text id="lisere1" x="160" y="15">sample3</text><line x1="95" y1="330" x2="565" y2="330" stroke="grey"/><line x1="100" y1="335" x2="100" y2="55" stroke="grey"/><text x="90" y="335" style="text-anchor:end;">0 %</text><line x1="95" y1="240.0" x2="565" y2="240.0" stroke="#B3B3B3"/><text x="90" y="245.0" style="text-anchor:end; baseline-shift:5;">10 %</text><line x1="95" y1="150.0" x2="565" y2="150.0" stroke="#B3B3B3"/><text x="90" y="155.0" style="text-anchor:end; baseline-shift:5;">20 %</text><line x1="95" y1="60.0" x2="565" y2="60.0" stroke="#B3B3B3"/><text x="90" y="65.0" style="text-anchor:end; baseline-shift:5;">30 %</text><line x1="157.5" y1="330" x2="157.5" y2="123.0" style=" stroke:grey; stroke-width:15 "/><text x="157.5" y="350" style="text-anchor:middle">ATM</text><g fill="grey"><text x="192.5" y="119.0" style="text-anchor:middle" >H111P</text></g><line x1="272.5" y1="330" x2="272.5" y2="276.0" style=" stroke:grey; stroke-width:15 "/><text x="272.5" y="350" style="text-anchor:middle">BIRC3</text><g fill="grey"><text x="307.5" y="272.0" style="text-anchor:middle" >R22P</text></g><line x1="387.5" y1="330" x2="387.5" y2="195.0" style=" stroke:grey; stroke-width:15 "/><text x="387.5" y="350" style="text-anchor:middle">TP53</text><g fill="grey"><text x="422.5" y="191.0" style="text-anchor:middle" >W40S</text></g><line x1="502.5" y1="330" x2="502.5" y2="60.0" style=" stroke:grey; stroke-width:15 "/><text x="502.5" y="350" style="text-anchor:middle">BRAF</text><g fill="grey"><text x="537.5" y="56.0" style="text-anchor:middle" >Y999Z</text></g></svg>'
svg_code4 = '<?xml version="1.0" encoding="utf-8"?> <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN" "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd"> <svg width="620px" height="420px" xml:lang="fr" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><title>My VAF</title><rect x="0" y="0" width="620" height="420" fill="white" opacity="0.7"/><path d="M 150,10 L 10,10 L 10,410 L 610,410 L 610,10 L 250,10" style="fill:none; stroke:grey;"/><text id="lisere1" x="160" y="15">sample3</text><line x1="95" y1="330" x2="565" y2="330" stroke="grey"/><line x1="100" y1="335" x2="100" y2="55" stroke="grey"/><text x="90" y="335" style="text-anchor:end;">0 %</text><line x1="95" y1="240.0" x2="565" y2="240.0" stroke="#B3B3B3"/><text x="90" y="245.0" style="text-anchor:end; baseline-shift:5;">10 %</text><line x1="95" y1="150.0" x2="565" y2="150.0" stroke="#B3B3B3"/><text x="90" y="155.0" style="text-anchor:end; baseline-shift:5;">20 %</text><line x1="95" y1="60.0" x2="565" y2="60.0" stroke="#B3B3B3"/><text x="90" y="65.0" style="text-anchor:end; baseline-shift:5;">30 %</text><line x1="157.5" y1="330" x2="157.5" y2="123.0" style=" stroke:grey; stroke-width:15 "/><text x="157.5" y="350" style="text-anchor:middle">ATM</text><g fill="grey"><text x="192.5" y="119.0" style="text-anchor:middle" >H111P</text></g><line x1="272.5" y1="330" x2="272.5" y2="276.0" style=" stroke:grey; stroke-width:15 "/><text x="272.5" y="350" style="text-anchor:middle">BIRC3</text><g fill="grey"><text x="307.5" y="272.0" style="text-anchor:middle" >R22P</text></g><line x1="387.5" y1="330" x2="387.5" y2="195.0" style=" stroke:grey; stroke-width:15 "/><text x="387.5" y="350" style="text-anchor:middle">TP53</text><g fill="grey"><text x="422.5" y="191.0" style="text-anchor:middle" >W40S</text></g><line x1="502.5" y1="330" x2="502.5" y2="60.0" style=" stroke:grey; stroke-width:15 "/><text x="502.5" y="350" style="text-anchor:middle">BRAF</text><g fill="grey"><text x="537.5" y="56.0" style="text-anchor:middle" >Y999Z</text></g></svg>'


def build_png(svg, png):
	fic = open(png,'wb')
	cairosvg.svg2png(bytestring=svg,write_to=fic)
	fic.close()

build_png(svg_code4, "glop.png")

# CAF 02 72 64 46 33


js/script.js

function test (form) {
	alert('TEST OK !');
}

function checkFile (form, file) {
	extOK = false ;
	extensionsOK = new Array('.txt', '', '.csv');						// ne pas oublier le point
	file = file.slice(file.indexOf("\\") + 1);
	ext = file.slice(file.lastIndexOf(".")).toLowerCase();
	//alert(extensionsOK.join(", "));
	for (var i = 0 ; i < extensionsOK.length; i++) {
		if ( extensionsOK[i] == ext ) { extOK = true ; }
	}

	if ( extOK ) { form.submit(); }
	
	else {
		alert ("Les types de fichiers acceptés sont : "
		+ extensionsOK.join(" ")
		+ ".\nMerci de télécharger un fichier au bon format.") ;
	}
}

function enforce_path (evt) {
	var path = evt.target ;
	var currentStrokeWidth = path.getAttribute("stroke-width")
	if (currentStrokeWidth == 4)
		path.setAttribute("stroke-width", "8")
	else
		path.setAttribute("stroke-width", "4")
	}




cgi-bin/myvaf.py
#!/usr/bin/env python3
#! -*- coding:utf-8 -*-

"""
Liens :
	http://webpython.codepoint.net/cgi_file_upload
	//// la réponse faites est très intérressante (WSGIScriptAlias)
	http://stackoverflow.com/questions/882430/how-to-hide-cgi-bin-py-etc-from-my-urls
"""

import os, sys
import cgi
import cgitb; cgitb.enable() 								# Affiche les erreurs dans le browser
# import cairosvg												# pour générer un graphique au format png
import common

form = cgi.FieldStorage()									# Récupère le contenu du formulaire

def file_check_ok(texte):
	"""
	Vérifier si la syntaxe du fichier est correcte
	En entrée : le texte à controller
	En sortie : vrai ou faux
	"""
	if texte == "":
		return False
	return True

def calcule_svg1(texte):
	"""
	Crée le fichier svg
	input : liste [ [ligne1, champs1], [ligne2, champs2], etc. ]
	output : String représentant le fichier SVG
	"""
	################# ICI DEFINIT LES POINTS DU SVG #####################
	largeur = 620											# largeur du graphique
	hauteur = 420											# hauteur du graphique
	# les abcisses
	abc1 = 100 ; abc2 = 560									# longueur utile de la barre des abcisses
	abc_t = abc2 - abc1										# longueur utile pour les abcisses
	eb = (abc2-abc1) / len(texte) 							# espace entre chaque barre (1/2 espace au début et à la fin)
	x_mid = largeur / 2										# axe vertical median
	lis_start = x_mid - len(fichier) * 6					# point d'abscisse de départ du liseré
	lis_stop = x_mid + len(fichier) * 6						# point d'abscisse d'arrivée du liseré
	# les ordonnées
	ord1 = 330 ; ord2 = 60									# ord1 : base ; ord2 : hauteur max
	ord_t = ord1 - ord2										# longueur utile pour les ordonnées
	# plus_grand : plus grande valeur de variant
	plus_grand = max([ int(v) for w,x,y,z,v in texte])		# plus_grande valeur trouvée (pour le calcul des traits d'ordonnées)
	# vsup : valeur max du graphique
	nb_dec = len(str(plus_grand))							# nombre de décimale de plus_grand
	if nb_dec == 1:											# si une seule décimale : vsup = 10
		vsup = 10
	elif plus_grand % 10 == 0:								# si valeur maxi est multiple de 10 alors
		vsup = plus_grand									# 	vsup = valeur maxi
	#elif int(str(plus_grand)[0]) == 1:						# si valeur max a sa unité de poids fort = 1
	#	 vsup = plus_grand - (plus_grand % 10)				# 	vsup = dizaine inférieure de la plus grande valeur
	else:													# dans les autres cas
		vsup = plus_grand - (plus_grand % 10) + 10			# 	vsup = dizaine supérieure de la plus grande valeur
	# déterminer les autres valeurs d'ordonnées du graphiques
	val_dec = int(str(vsup)[0])
	if vsup % 70 == 0:
		vsup = 8 * 10**(len(str(vsup))-1)					# on arrondi à 8 si on était à 7 (ex 70 --> 80)
	if vsup == 10:
		val_ord = (5, 10)
	elif vsup % 30 == 0:
		val_ord = (vsup // 3 , (vsup // 3) * 2 , vsup)
	elif vsup % 20 == 0 or vsup % 50 == 0:
		val_ord = (vsup // 2, vsup)
	# calculer l'espace entre les lignes des abcisses
	ea = (ord1 - ord2) / len(val_ord)	
	
	################# ICI ON COMMENCE LA DEFINITION DU SVG #####################
	svg = '<svg width="'+str(largeur)+'px" height="'+str(hauteur)+'px" xml:lang="fr" '
	svg += 'xmlns="http://www.w3.org/2000/svg" '
	svg += 'xmlns:xlink="http://www.w3.org/1999/xlink">'
	svg += '<title>My VAF</title>'
	
	
	# Cadre exterieur (en fonction de la définition de l'écran, avec un max et min)
	svg += '<rect x="0" y="0" width="620" height="420" fill="white"/>'
	# Position du liseré exterieur 
	svg += '<path d="M '+str(lis_start)+',10 L 10,10 L 10,410 L 610,410 L 610,10 L '+str(lis_stop)+',10" style="fill:none; stroke:grey;"/>'
	# texte du liseré
	svg += '<text id="lisere1" x="'+str(x_mid)+'" y="15" text-anchor="middle">' + fichier + '</text>'
	# ligne des abcisses
	svg += '<line x1="' + str(abc1-5) +'" y1="330" x2="' + str(abc2+5) + '" y2="330" stroke="grey"/>'
	# ligne de base des ordonnées
	svg += '<line x1="'+str(abc1)+'" y1="'+ str(ord1+5)+'" x2="'+str(abc1)+'" y2="'+str(ord2-5)+'" stroke="grey"/>'
	svg += '<text x="'+str(abc1-10) + '" y="'+str(ord1+5) + '" style="text-anchor:end;">0 %</text>'
	# lignes intermédiaires des ordonnées
	Y = ord1
	for a in val_ord:
		Y -= ea
		svg += '<line x1="'+str(abc1-5) + '" y1="'+str(Y) + '" x2="' + str(abc2+5) + '" y2="'+str(Y) + '" stroke="#B3B3B3"/>'
		svg += '<text x="'+str(abc1-10) + '" y="'+str(Y+5) + '" style="text-anchor:end; baseline-shift:5;">'+ str(a) +' %</text>'
	# Position des barres
	X = abc1
	for a in (texte):
		if X == abc1:									# compter 1/2 décalage pour la première barre 
			X = X + eb / 2
		else:											# sinon on décale de eb (ecart entre barre)
			X = X + eb
		Y2 = ord1 - (ord_t * int(a[4]) / vsup)
		svg += '<line x1="'+str(X) + '" y1="'+str(ord1)+'" x2="'+str(X) + '" y2="'+ str(Y2) + '" style=" stroke:grey; stroke-width:15 ">'
		svg += '<animate attributeName="y2" from="'+str(ord1)+'" to="'+str(Y2)+ '" begin="0s" dur="2s" />'
		svg += '</line>'
		svg += '<text x="'+str(X) + '" y="' + str(ord1 + 20) + '" style="text-anchor:middle">' + a[0] + '</text>'
		svg += '<ellipse cx="'+str(X+23)+'" cy="'+str(Y2-15)+'" rx="'+str(len(a[3])*5)+'" ry="14" style="fill:white; stroke-width:1; stroke:transparent"/>'
		#svg += '<set attributeName="fill" to="white" begin="2" />'
		#svg += '<set attributeName="stroke" to="grey" begin="2" /></ellipse>'
		svg += '<text x="'+str(X+23) + '" y="' + str(Y2-15) + '" style="text-anchor:middle; fill:grey; dominant-baseline:middle">' + a[3]
		svg += '</text>'
		#svg += '<set attributeName="fill" to="grey" begin="2" /></text>'

	################# ICI ON TERMINE LA DEFINITION DU SVG #####################
	svg += '</svg>'									# la fin du fichier SVG
	return svg										# il ne reste plus qu'à retourner le fichier

def debug(texte):
	"""
	Pour le debug
	"""
	print('<br/><h3>--- DEBUG ---</h3>')
	# print(texte)										# Afficher la liste 'texte'
	for i,line in enumerate(texte):						# afficher les éléments des éléments de 'texte'
		for j,field in enumerate(line):
			print("ligne {} - champs {} : {}<br/>".format(i, j, field))


def affiche_env():
	"""
	Affiche les variables renvoyées par Apache
	"""
	print("<br/><h3>--- VARIABLES D'ENVIRONNEMENT ---</h3>")
	for a,b in os.environ.items():
		print("{} -> {}<br/>".format(a,b))

#####################################################
#####         AFFICHAGE DE LA PAGE 				#####
#####################################################

print("""content-type: text/html\n
<!--#include virtual="/head.html" -->
<!--#include virtual="/header.html" -->
""")

print("""
<section>
<article>
<h1>MyVAF</h1>
<p>
	Affiche un graphique de la fréquence des allèles variants pour un échantillon.
</p>
<p>
	Un fichier d'exemple est téléchargeable : <a href="/static/sample-myvaf.txt" download="sample-myvaf.txt" id="upload_file" >sample-myvaf.txt</a>
</p>
<br />
""")

common.formulaire("myvaf.py")

if os.environ['REQUEST_METHOD'] == 'POST':		# si la méthode est 'POST'
		
	fileitem = form['fichier_csv']

	# définit le nom du fichier sans son extension	
	fichier = fileitem.filename
	for i,a in enumerate(fichier):
		if a == ".": ind = i
	fichier = fichier[:ind]
	# définit le nom et emplacement du graphique png
	png_path = "../graphics/" + fichier + ".png"
	png_file = fichier + ".png"
	# définit le nom et emplacement du graphique svg
	#svg_path = "../graphics/" + fichier + ".svg"
	#svg_file = fichier + ".svg"	
	print("<h4>{}</h4>".format(fileitem.filename))	# on affiche le nom du fichier
		
	texte = fileitem.file.read()				# on passe dans la variable texte la partie texte
	texte = common.file_format(texte)			# mettre le fichier en unicode
	if file_check_ok(texte):					# Si le controle du format de fichier est OK
		texte = common.parse_myvaf(texte)		# on analyse on fichier
		svg = calcule_svg1(texte)				# on crée le graphique svg1
		print(svg)								# AFFICHER LE SVG
		common.build_png_file(svg, png_path)	# créer un fichier PNG
		common.upload_png(png_path, png_file)	# créer le lien de téléchargement de ce PNG
		#debug(texte)							# pour debugguer
		#affiche_env()							# affiche les variables d'environnement renvoyées par Apache
		
	
print("""
</article>
</section>
<!--#include virtual="/footer.html" -->
""")


cgi-bin/index.py
#!/usr/bin/env python3
#! -*- coding:utf-8 -*-

"""
Liens :
	http://webpython.codepoint.net/cgi_file_upload
	//// la réponse faites est très intérressante (WSGIScriptAlias)
	http://stackoverflow.com/questions/882430/how-to-hide-cgi-bin-py-etc-from-my-urls
"""

import os, sys
import cgi
import cgitb; cgitb.enable() 								# Affiche les erreurs dans le browser
import cairosvg												# pour générer un graphique au format png
import common

form = cgi.FieldStorage()

def file_format(texte):										# mettre en unicode
	try:
		t = texte.decode("utf-8")
	except:
		t = texte.decode("ISO-8859-1")
	return t

def file_check_ok(texte):
	"""
	Vérifier si la syntaxe du fichier est correcte
	En entrée : le texte à controller
	En sortie : vrai ou faux
	"""
	if texte == "":
		return False
	return True

def calcule_svg1(texte):
	"""
	Crée le fichier svg
	input : liste [ [ligne1, champs1], [ligne2, champs2], etc. ]
	output : String représentant le fichier SVG
	"""
	################# ICI DEFINIT LES POINTS DU SVG #####################
	# les abcisses
	abc1 = 100 ; abc2 = 560									# longueur utile de la barre des abcisses
	abc_t = abc2 - abc1										# longueur utile pour les abcisses
	eb = (abc2-abc1) / len(texte) 							# espace entre chaque barre (1/2 espace au début et à la fin)
	# les ordonnées
	ord1 = 330 ; ord2 = 60									# ord1 : base ; ord2 : hauteur max
	ord_t = ord1 - ord2										# longueur utile pour les ordonnées
	# plus_grand : plus grande valeur de variant
	plus_grand = max([ int(v) for w,x,y,z,v in texte])		# plus_grande valeur trouvée (pour le calcul des traits d'ordonnées)
	# vsup : valeur max du graphique
	nb_dec = len(str(plus_grand))							# nombre de décimale de plus_grand
	if nb_dec == 1:											# si une seule décimale : vsup = 10
		vsup = 10
	elif plus_grand % 10 == 0:								# si valeur maxi est multiple de 10 alors
		vsup = plus_grand									# 	vsup = valeur maxi
	#elif int(str(plus_grand)[0]) == 1:						# si valeur max a sa unité de poids fort = 1
	#	 vsup = plus_grand - (plus_grand % 10)				# 	vsup = dizaine inférieure de la plus grande valeur
	else:													# dans les autres cas
		vsup = plus_grand - (plus_grand % 10) + 10			# 	vsup = dizaine supérieure de la plus grande valeur
	# déterminer les autres valeurs d'ordonnées du graphiques
	val_dec = int(str(vsup)[0])
	if vsup % 70 == 0:
		vsup = 8 * 10**(len(str(vsup))-1)					# on arrondi à 8 si on était à 7 (ex 70 --> 80)
	if vsup == 10:
		val_ord = (5, 10)
	elif vsup % 30 == 0:
		val_ord = (vsup // 3 , (vsup // 3) * 2 , vsup)
	elif vsup % 20 == 0 or vsup % 50 == 0:
		val_ord = (vsup // 2, vsup)
	# calculer l'espace entre les lignes des abcisses
	ea = (ord1 - ord2) / len(val_ord)	
	
	################# ICI ON COMMENCE LA DEFINITION DU SVG #####################
	svg = '<?xml version="1.0" encoding="utf-8"?> '
	svg += '<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN" '
	svg += '"http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd"> '
	svg += '<svg width="620px" height="420px" xml:lang="fr" '
	svg += 'xmlns="http://www.w3.org/2000/svg" '
	svg += 'xmlns:xlink="http://www.w3.org/1999/xlink">'
	svg += '<title>My VAF</title>'
	
	
	# Cadre exterieur (en fonction de la définition de l'écran, avec un max et min)
	svg += '<rect x="0" y="0" width="620" height="420" fill="white"/>'
	# Position du liseré exterieur 
	svg += '<path d="M 150,10 L 10,10 L 10,410 L 610,410 L 610,10 L 250,10" style="fill:none; stroke:grey;"/>'
	# texte du liseré
	svg += '<text id="lisere1" x="160" y="15">' + fichier + '</text>'
	# ligne des abcisses
	svg += '<line x1="' + str(abc1-5) +'" y1="330" x2="' + str(abc2+5) + '" y2="330" stroke="grey"/>'
	# ligne de base des ordonnées
	svg += '<line x1="'+str(abc1)+'" y1="'+ str(ord1+5)+'" x2="'+str(abc1)+'" y2="'+str(ord2-5)+'" stroke="grey"/>'
	svg += '<text x="'+str(abc1-10) + '" y="'+str(ord1+5) + '" style="text-anchor:end;">0 %</text>'
	# lignes intermédiaires des ordonnées
	Y = ord1
	for a in val_ord:
		Y -= ea
		svg += '<line x1="'+str(abc1-5) + '" y1="'+str(Y) + '" x2="' + str(abc2+5) + '" y2="'+str(Y) + '" stroke="#B3B3B3"/>'
		svg += '<text x="'+str(abc1-10) + '" y="'+str(Y+5) + '" style="text-anchor:end; baseline-shift:5;">'+ str(a) +' %</text>'
	# Position des barres
	X = abc1
	for a in (texte):
		if X == abc1:									# compter 1/2 décalage pour la première barre 
			X = X + eb / 2
		else:											# sinon on décale de eb (ecart entre barre)
			X = X + eb
		Y2 = ord1 - (ord_t * int(a[4]) / vsup)
		svg += '<line x1="'+str(X) + '" y1="'+str(ord1)+'" x2="'+str(X) + '" y2="'+ str(Y2) + '" style=" stroke:grey; stroke-width:15 ">'
		svg += '<animate attributeName="y2" from="'+str(ord1)+'" to="'+str(Y2)+ '" begin="0s" dur="2s" />'
		svg += '</line>'
		svg += '<text x="'+str(X) + '" y="' + str(ord1 + 20) + '" style="text-anchor:middle">' + a[0] + '</text>'
		svg += '<ellipse cx="'+str(X+23)+'" cy="'+str(Y2-15)+'" rx="'+str(len(a[3])*8)+'" ry="14" style="fill:white; stroke-width:1; stroke:transparent"/>'
		#svg += '<set attributeName="fill" to="white" begin="2" />'
		#svg += '<set attributeName="stroke" to="grey" begin="2" /></ellipse>'
		svg += '<text x="'+str(X+23) + '" y="' + str(Y2-15) + '" style="text-anchor:middle; fill:grey; dominant-baseline:middle">' + a[3]
		svg += '</text>'
		#svg += '<set attributeName="fill" to="grey" begin="2" /></text>'

	################# ICI ON TERMINE LA DEFINITION DU SVG #####################
	svg += '</svg>'									# la fin du fichier SVG
	return svg										# il ne reste plus qu'à retourner le fichier

def build_png_file(svg, png):
	"""
	Crée un fichier png pour le téléchargement
	"""	
	#png = cairosvg.svg2png(bytestring = svg)
	cairosvg.svg2png(bytestring=bytes(svg,'UTF-8'), write_to = png)

def upload_png(png):
	"""
	Affiche le lien de téléchargement du fichier png
	"""
	# print("<p>"+png_file+"</p>")
	print('<br /><a href="' + png + '" download="' + png_file + '" id="upload_file" >Télécharger le graphique<a>' )


def build_svg_file(svgContent, svgFile):
	"""
	Crée un fichier svg pour le téléchargement
	"""
	with open(svgFile, 'w') as f:
		f.write(svg)

def upload_svg(svg):
	"""
	Affiche le lien de téléchargement du fichier svg
	"""
	# print("<p>"+png_file+"</p>")
	print('<br /><a href="' + svg + '" download="' + svg_file + '" id="upload_file" >Télécharger le graphique<a>' )


def debug(texte):
	"""
	Pour le debug
	"""
	print('<br/><h3>--- DEBUG ---</h3>')
	# print(texte)										# Afficher la liste 'texte'
	for i,line in enumerate(texte):						# afficher les éléments des éléments de 'texte'
		for j,field in enumerate(line):
			print("ligne {} - champs {} : {}<br/>".format(i, j, field))


def affiche_env():
	"""
	Affiche les variables renvoyées par Apache
	"""
	print("<br/><h3>--- VARIABLES D'ENVIRONNEMENT ---</h3>")
	for a,b in os.environ.items():
		print("{} -> {}<br/>".format(a,b))

#####################################################
#####         AFFICHAGE DE LA PAGE 				#####
#####################################################

print("""content-type: text/html\n
<!--#include virtual="/head.html" -->
<!--#include virtual="/header.html" -->
""")

print("""
<section>
<article>
<h1>MyVAF</h1>
<p>
	Affiche un graphique de la fréquence des allèles variants pour un échantilon.<br/>
	Un fichier d'exemple est téléchargeable : <a href="/static/sample-myvaf.txt" download="sample-myvaf.txt" id="upload_file" >sample-myvaf.txt</a>
</p>
<br />
""")

common.formulaire("index.py")

#formulaire()

if os.environ['REQUEST_METHOD'] == 'POST':		# si la méthode est 'POST'
		
	fileitem = form['fichier_csv']

	# définit le nom du fichier sans son extension	
	fichier = fileitem.filename
	for i,a in enumerate(fichier):
		if a == ".": ind = i
	fichier = fichier[:ind]
	# définit le nom et emplacement du graphique png
	png_path = "../graphics/" + fichier + ".png"
	png_file = fichier + ".png"
	# définit le nom et emplacement du graphique svg
	svg_path = "../graphics/" + fichier + ".svg"
	svg_file = fichier + ".svg"	
	print("<h4>{}</h4>".format(fileitem.filename))	# on affiche le nom du fichier
		
	texte = fileitem.file.read()				# on passe dans la variable texte la partie texte
	texte = file_format(texte)					# mettre le fichier en unicode
	if file_check_ok(texte):					# Si le controle du format de fichier est OK
		texte = common.parse_myvaf(texte)		# on analyse on fichier
		svg = calcule_svg1(texte)				# on crée le fichier svg1
		print(svg)								# AFFICHER LE SVG
		common.build_png_file(svg, png_path)	# créer un fichier PNG (pas au point)
		common.upload_png(png_path, png_file)	# créer le lien de téléchargement de ce PNG
		#build_svg_file(svg, svg_path)			# créer un fichier SVG
		#upload_svg(svg_path)					# créer le lien de téléchargement de ce SVG
		#debug(texte)							# pour debugguer
		#affiche_env()							# affiche les variables d'environnement renvoyées par Apache
		
	
print("""
</article>
</section>
<!--#include virtual="/footer.html" -->
""")
